### PR TITLE
2828  event tlm database

### PR DESCRIPTION
**Pull request by grignards**
Refs #2828 

## Checkpoints

The goal of this section is to help code reviewers and the merge request author to do the required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present
- [x] Automated tests have been written (covering the feature or non regression), or *at least* an issue has been created to implement the test

**For developers**

- [x] Manual tests done to ensure stability on whole application layers, issue expectation follow
- [x] The original issue has been updated with a short summary of the implemented solution

**For the quality team**

- [x] Tested by the quality team

### CODE REVIEW

File: backend/tracim_backend/lib/core/user.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def get_all(self) -> typing.Iterable[User]:
 
     def get_user_ids_from_profile(self, profile: Profile) -> typing.Iterable[int]:
         query = self._apply_base_filters(self._session.query(User.user_id))
-        return query.filter(User.profile == profile).all()
+        return [res[0] for res in query.filter(User.profile == profile)]
```
Comment (inkhey) [inline backend/tracim_backend/lib/core/user.py:151]: Not sure to understand the meaning of this change. Can you explain @grignards ?

Comment (grignards) [reply backend/tracim_backend/lib/core/user.py:151]: SQLAlchemy returns tuples, even when only one column is asked for (like it is here), so I extract the ids from their tuples.

Comment (inkhey) [reply backend/tracim_backend/lib/core/user.py:151]: ok

File: backend/tracim_backend/command/__init__.py [hunk 2]
```diff
@@ -63,6 +63,9 @@ def take_action(self, parsed_args: argparse.Namespace) -> None:
                         self.take_app_action(parsed_args, app_context)
         except Exception as exc:
             logger.exception(self, exc)
+            import traceback
```
Comment (inkhey) [inline backend/tracim_backend/command/__init__.py:66]: logger.exception(self, exc) is not enought to get traceback ?

File: backend/tracim_backend/lib/event.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            receiver_ids.add(event.user["user_id"])
+        return receiver_ids
+
+    @classmethod
+    def _get_workspace_event_receiver_ids(cls, event: Event) -> typing.Set[int]:
+        user_api = UserApi(current_user=None, session=cls._session(), config=cls._config())
+        administrators = user_api.get_user_ids_from_profile(Profile.ADMIN)
+        role_api = RoleApi(current_user=None, session=cls._session(), config=cls._config())
+        workspace_members = role_api.get_workspace_member_ids(event.workspace["workspace_id"])
+        return set(administrators).union(set(workspace_members))
```
Comment (raphj) [inline backend/tracim_backend/lib/event.py:287]:
```suggestion
        return set(administrators).union(workspace_members)
```

```suggestion
        return set(administrators + workspace_members)
```
(works with Python lists, not sure about the types of `administrators` and `workspace_members`)

File: backend/tracim_backend/lib/event.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        self._current_user = None
+
+    # User events
+    @hookimpl
+    def on_user_created(self, user: User, db_session: TracimSession) -> None:
+        self._create_user_event(OperationType.CREATED, user, db_session)
+
+    @hookimpl
+    def on_user_modified(self, user: User, db_session: TracimSession) -> None:
+        if user.is_deleted:
```
Comment (raphj) [inline backend/tracim_backend/lib/event.py:98]: Are we sure there aren't edge cases where a deleted user is modified, making this function behave erratically (if I understand its logic correctly)? 

Same remark for all these functions using `something.is_deleted`

File: backend/tracim_backend/command/__init__.py (hunk 2 above)
Comment (grignards) [reply backend/tracim_backend/command/__init__.py:66]: the logs do not appear in the tests, that's why I added this and I forgot it. I didn't check if logs are displayed while using `tracimcli` interactively.

File: backend/tracim_backend/models/event.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+"""
+Live messages/events related model.
+
+It's perfectly fine to re-use this definition in the tracim application,
+though.
```
[the remark this answers is not in the document]
Comment (grignards) [reply backend/tracim_backend/models/event.py:6]: Copy/paste… I'll remove this comment as it does not make sense here.

File: backend/tracim_backend/lib/event.py (hunk 4 above)
Comment (inkhey) [reply backend/tracim_backend/lib/event.py:98]: @grignards , is there any way to have some kind of diff between previews and new User at this level ?

Comment (grignards) [reply backend/tracim_backend/lib/event.py:98]: Yes it is possible through [the attribute state](https://docs.sqlalchemy.org/en/13/orm/internals.html#sqlalchemy.orm.state.AttributeState.history)

Comment (grignards) [reply backend/tracim_backend/lib/event.py]: Attribute state works as expected, I modified my code to ensure behavior (plus a unit test for the corner case of is_deleted==True then modified).

File: backend/tracim_backend/lib/event.py (hunk 3 above)
Comment (grignards) [reply backend/tracim_backend/lib/event.py]: I didn't know that `union` worked with a list as parameter, thank you for the tip.

File: backend/tracim_backend/lib/event.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        query = (
+            self._session.query(Message)
+            .filter(Message.receiver_id == user_id)
+            .options(joinedload(Message.event))
+        )
+        if read_status == ReadStatus.READ:
+            query = query.filter(Message.read != null())
+        elif read_status == ReadStatus.UNREAD:
+            query = query.filter(Message.read == null())
+        return query.all()
```
[the remark this answers is not in the document]
Comment (grignards) [reply backend/tracim_backend/lib/event.py]: Else means `ALL` here and `ALL` means no filtering (so no action to take on the query).
I added a comment to explain it (plus an assert).

**Comment by inkhey** [comment]
Found a  probably 3rd issue:
- [ ] standard http message endpoint is not paginated, need to discuss with frontend, but this is probably needed. Don't think this is a blocking point but a specific issue is needed, if need is confirmed by frontend (@Skylsmoi)
